rust/agama-cli/src/auth.rs

```diff
@@ -0,0 +1,246 @@
+use clap::{arg, Args, Subcommand};
+use home;
+use reqwest::header::{HeaderMap, HeaderValue, CONTENT_TYPE};
+use std::fs;
+use std::fs::File;
+use std::io;
+use std::io::{BufRead, BufReader};
+use std::os::unix::fs::PermissionsExt;
+use std::path::{Path, PathBuf};
+
+const DEFAULT_JWT_FILE: &str = ".agama/agama-jwt";
+const DEFAULT_AUTH_URL: &str = "http://localhost:3000/authenticate";
+const DEFAULT_FILE_MODE: u32 = 0o600;
+
+#[derive(Subcommand, Debug)]
+pub enum AuthCommands {
+    /// Login with defined server. Result is JWT stored locally and made available to
+    /// further use. Password can be provided by commandline option, from a file or it fallbacks
+    /// into an interactive prompt.
+    Login(LoginArgs),
+    /// Release currently stored JWT
+    Logout,
+    /// Prints currently stored JWT to stdout
+    Show,
+}
+
+/// Main entry point called from agama CLI main loop
+pub async fn run(subcommand: AuthCommands) -> anyhow::Result<()> {
+    match subcommand {
+        AuthCommands::Login(options) => login(LoginArgs::proceed(options).password()?).await,
+        AuthCommands::Logout => logout(),
+        AuthCommands::Show => show(),
+    }
+}
+
+/// Reads stored JWT token and returns it
+pub fn jwt() -> anyhow::Result<String> {
+    if let Some(file) = jwt_file() {
+        if let Ok(token) = read_line_from_file(&file.as_path()) {
+            return Ok(token);
+        }
+    }
+
+    return Err(anyhow::anyhow!("JWT not available"));
+}
+
+/// Stores user provided configuration for login command
+#[derive(Args, Debug)]
+pub struct LoginArgs {
+    #[arg(long, short = 'p')]
+    password: Option<String>,
+    #[arg(long, short = 'f')]
+    file: Option<PathBuf>,
+}
+
+impl LoginArgs {
+    /// Transforms user provided options into internal representation
+    /// See Credentials trait
+    fn proceed(options: LoginArgs) -> Box<dyn Credentials> {
```

Comment:
```rust
    fn proceed(options: LoginArgs) -> Box<dyn Credentials> {
        if let Some(password) = options.password {
            Box::new(KnownCredentials { password })
        } else if let Some(path) = options.file {
            Box::new(FileCredentials { path })
        } else {
            Box::new(MissingCredentials {})
        }
    }
```

Reply:
ah yes ... I was surprised that the match passed through the first round ;-)

I really don't know if the ```if ... else``` version is much better, but will adapt.